Migration docs (DCNE-321)

akinross opened the pull request.

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+subcategory: "Guides"
+page_title: "Migration"
+description: |-
+    An overview of resource and datasource migration changes
```
Can we add migration from Terraform SDKv2 to Plugin Framework?

shrsr replied:
or just migration to the latest Terraform Plugin Framework?

akinross replied:
removed latest but went with your second suggestion and also added to title

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+
+## Introduction
+
+Since the first release in September 2020, the Terraform ACI provider has come a long way, adding new resources and datasources to make managing Cisco ACI environments easier. Over the years, we've moved from SDKv1 to [SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2), and now to the latest [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework), whic ...
```
Since the first release -> Since its first release

shrsr replied:
datasources -> data sources ?

akinross replied:
changed everywhere

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+
+Before making any changes, it's crucial to back up your current Terraform state. This ensures that you can restore your environment if anything goes wrong during the upgrade process.
+
+1. **Local Backend**: Navigate to Your Terraform Project Directory and open a terminal and navigate to the directory containing your Terraform configuration files and copy the current state file (`terraform.tfstate`) to a backup location.
```
Navigate to Your Terraform Project Directory and open a terminal and navigate to the directory containing your Terraform configuration files and copy the current state file (`terraform.tfstate`) to a backup location.
->
Open a terminal and navigate to the directory containing your ACI Terraform configuration files with the state and copy the current state file (`terraform.tfstate`) to a backup location.

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+2. **Enhance Consistency**: Standardize attribute names and structures across resources and datasources.
+3. **Simplify Usage**: Remove outdated attributes to streamline the configuration process.
+
+We understand the importance of maintaining backward compatibility to avoid disrupting existing configurations. This is why during a transition period, both old and new style attributes can be used. It allows users to continue using their existing configurations while gradually adopting the new style attributes. We have implemented deprecation warnings for old style attributes, informing users of ...
```
new style attributes -> appropriately redefined attributes?

akinross replied:
closing this one and commenting on your comment below

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+
+We understand the importance of maintaining backward compatibility to avoid disrupting existing configurations. This is why during a transition period, both old and new style attributes can be used. It allows users to continue using their existing configurations while gradually adopting the new style attributes. We have implemented deprecation warnings for old style attributes, informing users of ...
+
+> For the same ACI property, the old and new style attributes cannot be used together. Doing so will cause the provider to error during configuration validation.
```
Would be good to change this "new style attributes"

akinross replied:
Yes think we as the team should decide on how to name this, have no preference will change accordingly everywhere when decision is made. 

@sajagana @samiib @anvitha-jain @gmicol @lhercot , any suggestions?

shrsr replied:
Example:
![Screenshot 2025-03-25 at 10 27 17 AM](https://github.com/user-attachments/assets/247a0fb9-6222-4f28-ae12-11571d0c93bf)

sajagana replied:
Can we use "legacy and modern/new attributes" to replace the "old and new style attributes"?

akinross replied:
putting suggestion of @shrsr here since this was in separate comment new style attributes -> appropriately redefined attributes?

akinross replied:
changed

shrsr commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+
+### Step 6: Cleanup
+
+1. **Remove Backup (Optional)**: If everything is working correctly, you can remove the backup state file.
```
Does the colour of the font change? There are a few inconsistencies

akinross replied:
What do you mean by this? The colours displayed here in GitHub? 

I used the https://registry.terraform.io/tools/doc-preview tool and VsCode, here all the titles and bullet points seems consistent

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,223 @@
+
+## Changed Annotation Behavior
+
+The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at the provider level.
```
added an example below

sajagana commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,232 @@
+
+Upgrading the ACI Terraform provider to a new version involves several steps to ensure a smooth transition and to avoid any disruptions to your existing configurations. This guide will walk you through the process, including taking a backup of your current state.
+
+#### Step 1: Backup Your Current State
```
```suggestion
### Step 1: Backup Your Current State
```

shrsr reviewed (dismissed):
Apart from the use of "new style attributes".. everything LGTM!

gmicol commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,232 @@
+
+## Error for Existing MO on Create
+
+In migrated resources, we provide the provider level option (`allow_existing_on_create`) to check during the plan if an object already exists. By default, we allow existing MOs to be managed, but this option can be set to `false` to allow for a safeguard mechanism to prevent managing the same MO by different configuration parts. The drawback of this safeguard mechanism is that during a plan, an a ...
```
```suggestion
Migrated resources offer a provider-level option (`allow_existing_on_create`) to check during the plan if an object already exists. By default, existing MOs can be managed, but this option can be set to `false`, providing a safeguard mechanism to prevent managing the same MO by different configuration parts. The drawback of this safeguard mechanism is an additional API call per resource during the plan to verify that the object does not already exist.
```

gmicol commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,232 @@
+
+## Allowing Empty Input for Attributes
+
+Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. The previous SDKs did not support null and unknown value states, which meant that before the Plugin Framework, we couldn't differentiate between an empty value and a value that wasn't provided. Because of this limitation, we were unable to update a string value to an empty string.
```
```suggestion
Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string was not possible before the Plugin Framework.
```

shrsr reviewed (approved):
Apart from the use of "new style attributes".. everything LGTM

gmicol reviewed (approved):
Maybe we have to revisit the "old/new style attributes" formulation but otherwise LGTM

akinross commented:
rebased to include ap migration

open discussion item old/new style attributes withs suggestions thus far:

- mixing both @sajagana and @shrsr solutions: legacy and redefined attributes
- @gaspard: original and updated attributes
- @sajagana: legacy and modern/new attributes
- @shreyas: new style attributes -> appropriately redefined attributes, no suggestion for old style

shrsr commented:
legacy and redefined attributes sound good to me

shrsr reviewed (dismissed):
LGTM

gmicol reviewed (approved):
LGTM

sajagana reviewed (dismissed):
LGTM!

sajagana reviewed (approved):
LGTM!

shrsr reviewed (approved):
LGTM

gmicol reviewed (approved):
LGTM

sajagana reviewed (approved):
LGTM!

gmicol reviewed (approved):
LGTM

shrsr commented on docs/guides/migration.md:
````diff
@@ -137,9 +137,74 @@ Before making any changes, it's crucial to back up your current Terraform state.
+    Successfully removed 1 resource instance(s).
+    ```
+
+    The `terraform plan` output should reflect no changes, when the resource is unmanaged.
````
Wouldn't terraform plan try to add the child resource resource back at this point?
Shall we mention that the child resource can be removed from the config file, once the resource is no longer managed?

shrsr replied:
Resolving as discussed

shrsr reviewed (approved):
LGTM

anvitha-jain reviewed (dismissed):
LGTM

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,297 @@
+
+> It is important to note that for the same ACI property, legacy and redefined attributes cannot be used simultaneously. Attempting to do so will result in an error during configuration validation.
+
+A downside to this approach is increased verbosity in the plan output due to known after applies for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```
changed a bit different from your suggestion, please press resolve if this is ok for you

samiib commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+> It is important to note that for the same ACI property, legacy and redefined attributes cannot be used simultaneously. Attempting to do so will result in an error during configuration validation.
+
+A downside to this approach is increased verbosity in the plan output due to "known after applies" for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```
```suggestion
A downside to this approach is increased verbosity in the plan output due to the "known after apply" state for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```

shrsr reviewed (dismissed):
LGTM

samiib reviewed (dismissed):
LGTM

sajagana reviewed (approved):
LGTM!

gmicol reviewed (approved):
LGTM

lhercot commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,315 @@
+    }
+    ```
+
+    !> Child resources should not be used in combination with nested attributes in parent resources. Doing so will result in unexpected behaviour.
````
I think we need to add a sentence here that is clearer like: Use one or the other, Either use a child or add it in the parent or something similar.

akinross replied:
ok will also add a note for the redesign to rephrase, because this sentence was previously defined and used already in resource docs

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+    !> Child resources should not be used in combination with nested attributes in parent resources. Doing so will result in unexpected behaviour.
+
+    In this scenario the child configuration should not be destroyed, but should be unmanaged (removed from state) before executing the plan. Terraform will display destroy intent in the `terraform plan` output when executed without unmanaging the child resource first.
```
```suggestion
    In this scenario the child configuration should not be destroyed, but should be unmanaged (removed from state) before executing the plan. Terraform will display destroy intent in the `terraform plan` output when executed without unmanaging the child resource first.
```
What is the scenario? Maybe describe the scenario before showing the example or rephrase this. Also, add a sentence or link to the next part when you show how to unmanage.

lhercot commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,315 @@
+    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
+    ```
+
+    The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in apic, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.has ...
````
```suggestion
    The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in APIC, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.hashicorp.com/terraform/cli/commands/plan#planning-modes) documentation of Terraform.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+> It is important to note that for the same ACI property, legacy and redefined attributes cannot be used simultaneously. Attempting to do so will result in an error during configuration validation.
+
+A downside to this approach is increased verbosity in the plan output due to the "known after apply" state for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```
```suggestion
A downside to this approach is the increased verbosity in the plan output when a change is detected due to the display of the "known after apply" state for each non-provided legacy attribute. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+In an ACI setup, Managed Objects (MOs) represent the different physical and logical parts within the Management Information Tree (MIT). These MOs can be linked through relationship MOs, which define how different MOs are connected. Relationship MOs act as connectors that establish links between two or more MOs, helping to organize and structure the hierarchical relationships within the MIT. In AC ...
+
+1. **Explicit relations** These require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.
```
```suggestion
1. **Explicit relationships** These relationships require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+In an ACI setup, Managed Objects (MOs) represent the different physical and logical parts within the Management Information Tree (MIT). These MOs can be linked through relationship MOs, which define how different MOs are connected. Relationship MOs act as connectors that establish links between two or more MOs, helping to organize and structure the hierarchical relationships within the MIT. In AC ...
+
+1. **Explicit relations** These require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.
+2. **Named relations** These require the name (identifier) attribute to be specified in the relationship MO, triggering a resolving mechanism based on precedence order. This allows the relationship to form with any MO in the precedence order.
```
```suggestion
2. **Named relationships** These relationships require the name (identifier) attribute to be specified in the relationship MO, triggering a resolving mechanism based on a precedence order. This allows the relationship to form with the first MO that exists in the precedence order.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+1. **Explicit relations** These require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.
+2. **Named relations** These require the name (identifier) attribute to be specified in the relationship MO, triggering a resolving mechanism based on precedence order. This allows the relationship to form with any MO in the precedence order.
+
+In non-migrated resources of the Terraform Provider ACI, the relationship types are hidden from the user by allowing a DN (resource ID) or name input. The relationship type determines which attribute (name vs tDn) is added to the payload. In SDKv2, there was no enforcement (only warnings in logs) of the final plan needing to match the applied state, which allowed us to strip the name from a provi ...
```
```suggestion
In non-migrated resources of the Terraform Provider ACI, the relationship types are hidden from the user by allowing a DN (resource ID) or name input. The relationship type determines which attribute (name vs tDn) is added to the payload. In SDKv2, there was no enforcement (only warnings in logs) of the final plan needing to match the applied state, which allowed us to extract the name from a provided DN for named relationships and use that extracted name in the relationship payload. However, the Plugin Framework enforces that the final plan values have to match the applied state.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+In non-migrated resources of the Terraform Provider ACI, the relationship types are hidden from the user by allowing a DN (resource ID) or name input. The relationship type determines which attribute (name vs tDn) is added to the payload. In SDKv2, there was no enforcement (only warnings in logs) of the final plan needing to match the applied state, which allowed us to strip the name from a provi ...
+
+Consequently, for the ACI Provider, using the name of the provided tDn for named relation MOs means the DN input may not match the resolved tDn, potentially causing provider errors. Due to this change, we decided to expose only configurable attributes as input for resources, requiring the name attribute instead of the DN for named relationships.
```
```suggestion
Consequently, for the ACI Provider, using the name of the provided tDn for the named relationship MOs means that the user provided DN may not match the resolved tDn, potentially causing provider errors. Due to this change, we decided to expose only configurable attributes as input for resources, requiring the name attribute instead of the DN for named relationships.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+Consequently, for the ACI Provider, using the name of the provided tDn for named relation MOs means the DN input may not match the resolved tDn, potentially causing provider errors. Due to this change, we decided to expose only configurable attributes as input for resources, requiring the name attribute instead of the DN for named relationships.
+
+To ensure the final plan matches the applied state, legacy attributes representing a Distinguished Name (DN) of a named relation must be resolved into the correct target Distinguished Name (tDn). This requires the object to exist when establishing the relationship; failure to do so will cause the provider to panic. See provider panic exmaple below.
```
```suggestion
To ensure the final plan matches the applied state, legacy attributes representing a Distinguished Name (DN) of a named relationship must be resolved into the correct target Distinguished Name (tDn). This requires the object to exist when establishing the relationship; the absence of the object will cause the provider to panic. See provider panic example below.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+## Allowing Empty Input for Attributes
+
+Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string was not possible before the Plugin Framework.
+
```
I would add a clear statement that it is now supported. It currently let the user assume it is supported but we should be clearer.

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+8. UI location information is provided for each configuration option.
+9. Resource DN format (resource ID) is documented.
+10. Both minimal and full examples are provided, featuring up to two parent resources when applicable.
+11. Examples for importing resources using both CLI commands and HCL configuration blocks are provided.
```
will add

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,315 @@
+
+## Changed Annotation Behavior
+
+The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at the provider level.
```
added an example for where you can overwrite the annotation? not sure what we should add as additional explanation

akinross commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,315 @@
+
+    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
+    ```
+
````
adjusted the sentence above

gmicol reviewed (dismissed):
LGTM

lhercot commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,339 @@
+   ***Deprecated Terraform version constraint format***
+   ```hcl
+   provider "aci" {
+     version = "x.y.z"  # Replace with the new version number
````
```suggestion
     version = "~> x.y.z"  # Replace with the new version number
```

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,339 @@
+2. **Remote Backend:** If you are using a remote backend (e.g., S3, Azure Blob Storage), ensure you have a backup of the state file from the remote location.
+
+### Step 2: Update the Provider Version
+
```
added

akinross commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,315 @@
+
+   ```bash
+   terraform init
+   ```
````
checked and changed the command to include -upgrade

anvitha-jain reviewed (approved):
LGTM

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,341 @@
+
+## Error for Existing MO on Create
+
+Migrated resources support the (`allow_existing_on_create`) provider-level option to check during the plan if an object already exists. By default, existing MOs will be automatically managed on apply, as (`allow_existing_on_create`) is set to `true` by default.  This option can be set to `false`, providing a mechanism to prevent the management of the same MO by separate configurations. The drawba ...
```
```suggestion
Migrated resources support the [`allow_existing_on_create`](https://registry.terraform.io/providers/CiscoDevNet/aci/latest/docs#allow_existing_on_create-1) provider-level option to check during the plan if an object already exists. By default, existing MOs will be automatically managed on apply, as `allow_existing_on_create` is set to `true` by default.  This option can be set to `false`, providing a mechanism to prevent the management of the same MO by separate configurations. The drawback of this mechanism is that it requires an additional API call per resource during the plan to verify that the object does not already exist.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,341 @@
+
+## Allowing Empty Input for Attributes
+
+Migrated resources in the Terraform ACI provider support null, unknown, and known states. Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty stri ...
```
```suggestion
The attributes of migrated resources in the Terraform ACI provider now support null, unknown, and known states which can be used to reset values to empty or to enforce that children objects are not present. Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string, updating a map to an empty map or a list of maps to an empty list was not possible before the Plugin Framework but is now supported in the migrated resources.
```

lhercot commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,341 @@
+
+Since its first release in September 2020, the Terraform ACI provider has significantly progressed, adding new resources and data sources to streamline the management of Cisco ACI environments. Over the years, we've transitioned from SDKv1 to [SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2), and now to the latest [Terraform Plugin Framework](https://developer.hashicorp.com/terrafor ...
+
+In order to fully leverage the new features of the Terraform Plugin Framework, we undertook a complete rewrite of the Terraform ACI provider's resources and data sources. Although this was a significant effort, it presented an opportunity to review and enhance the existing ACI provider. This guide provides detailed steps for migrating from previous versions, outlines key optimization steps to enh ...
```
```suggestion
In order to fully leverage the new features of the Terraform Plugin Framework, we decided to undertake a complete rewrite of the Terraform ACI provider's resources and data sources. Although this is a significant effort, it presented an opportunity to review and enhance the existing ACI provider and how the provider is developed. 

This complete rewrite will stretch a series of releases starting with [v2.16.0](https://github.com/CiscoDevNet/terraform-provider-aci/releases/tag/v2.16.0) where all new resources and data sources will be implemented using the new Terraform Plugin Framework and pre-existing resources and data sources will be slowly migrated into the Terraform Plugin Framework. Migrated resources and data sources contain a note that specifies it has been migrated at the top of their documentation. Once all pre-existing resources are migrated to the Terraform Provider Framework, a major release of the provider will be created to remove legacy attributes.

This guide provides detailed steps for migrating from previous versions, outlines key optimization steps to enhance performance, and aims to facilitate a smoother transition for users adapting to these updates. Additionally, it outlines how the migrated resources use the new features introduced by the Terraform Plugin Framework, along with the modifications we implemented in the ACI provider and the rationale behind these changes.
```

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,341 @@
+
+## Introduction
+
+Since its first release in September 2020, the Terraform ACI provider has significantly progressed, adding new resources and data sources to streamline the management of Cisco ACI environments. Over the years, we've transitioned from SDKv1 to [SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2), and now to the latest [Terraform Plugin Framework](https://developer.hashicorp.com/terrafor ...
```
ok

akinross commented on docs/guides/migration.md:
````diff
@@ -0,0 +1,341 @@
+
+1. **Local Backend:** Open a terminal and navigate to the directory where your ACI Terraform configuration files with the state are located. Copy the current state file (terraform.tfstate) to a backup location.
+
+   ```bash
+   cd /path/to/your/terraform/project
+   cp terraform.tfstate /path/to/your/backup/terraform.tfstate.backup
+   ```
````
ok fix all indentation

akinross commented on docs/guides/migration.md:
```diff
@@ -0,0 +1,341 @@
+
+    The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in APIC, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.has ...
+
+## Changes to the ACI Provider
```
changed let me know if this better

akinross replied:
undid change after discussion

gmicol reviewed (approved):
LGTM

lhercot reviewed (approved):
LGTM
